### PR TITLE
Add ability to save "local" help only

### LINKED ISSUE #2356: scons -h should print new options
This issue was originally created at: 2009-02-24 13:54:13.
This issue was reported by: `stevenknight`.

>stevenknight said at 2009-02-24 13:54:13

If you add options to SCons via `AddOption()`, SCons will print them in its optparse-style help shown by `scons -h`.
 
However, if you also add help text via `Help()`, SCons no longer prints the optparse-style help.  You can still see help on SCons's built-in command line options using `scons -H`, but since that doesn't parse the SConscript files it doesn't show any of the added options.
 
The feature request is for SCons to maintain a second optparse instance containing only the options added via `AddOption()`.  If help text has been added via `Help()`, SCons could print only the optparse help for this second optparse instance, followed by the text added via `Help()`.
 
The text of the final line of help: 'Use scons -H for help about command-line options.' should also be changed to something like  'Use scons -H for help about scons built-in command-line options.' since it doesn't print help about ALL command line options, only those built into SCons.

>gregnoel said at 2009-03-05 15:16:13

Bug party triage.

>stevenknight said at 2009-11-10 18:00:20

stevenknight => issues@scons

### LINKED ISSUE #3686: Don't lose local-options help text
This is related to other recent activity (issue on help, plus doc updating activities) but wanted to document the case specifically rather than hijacking a different issue.

Help message handling is logically awkward:

1. scons' own help text is always available with `-H`. It also provides the _initial_ text used for the local help, which is displayed with the `-h` option, which is the same as `--help`.
2. `AddOption` calls add to the local help, but not to the SCons help.
3. First call to `Help` either rewrites the local help, or appends to it, depending on value of `append` kwarg
5. Subsequent call to `Help` appends to local help, regardless of value of `append`
6. the `Variables` object method `GenerateHelpText` requires an `env` to operate on but `Variables` is not environment-specific.  Why is this?

Complaints:

- If the first `Help` call did not append, the help text from `AddOption` is lost: it's dropped from local help, and was not added to SCons help.
- Inconsistent interpretation of `Help(msg, append=bool)`: caller-supplied `append` honored on first call, ignored otherwise; default value of `append` appears to flip, first call defaults non-appending, subsequent calls default appending
- It should be possible to construct a local help message consisting of local options (`AddOption`) and local variables (aka: "help for project-specific command line usage"). It isn't, unless full SCons help is retained in the local help as well. That can well be too noisy for a project that adds a number of its own options.
- Help generated by `AddOption` calls gets a section banner (**Local Options**) but command-line variables do not.

### PR DESCRIPTION
**Pull request by mwichmann**
An additional keyword argument, `local_only`, is now recorgnized by `Help()`. If true, and `append` is `True`, then only project-defined help messages are saved into the help text. That is, save help added by calls to `AddOption`, but not SCons' own help, which is added a different way.

Fixes #2356
Fixes #3686

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

File: doc/user/output.xml [hunk 1]
```diff
@@ -120,11 +101,12 @@ Type: 'scons program' to build the production program,
...
 
     The &SConscript; files may contain
-    multiple calls to the &Help; function,
+    multiple calls to the &f-link-Help; function,
     in which case the specified text(s)
     will be concatenated when displayed.
-    This allows you to split up the
-    help text across multiple &SConscript; files.
+    This allows you to define bits of help text together with
+    the corresponding feature, even if speread
```
Comment (bdbaddog) [inline doc/user/output.xml]: typo `speread`.

File: doc/user/output.xml [hunk 2]
```diff
@@ -135,8 +117,16 @@ Type: 'scons program' to build the production program,
 
     <para>
 
-    When used with &AddOption; Help("text", append=False) will clobber any help output associated with AddOption().
-    To preserve the help output from AddOption(), set append=True.
+    Calling <literal>Help("text")</literal> will clobber
```
Comment (bdbaddog) [inline doc/user/output.xml]: maybe `overwrite` instead of `clobber` ?

File: doc/user/output.xml [hunk 3]
```diff
@@ -135,8 +117,16 @@ Type: 'scons program' to build the production program,
...
+    any help output associated with &f-link-AddOption; calls.
+    To preserve the help output from &AddOption;,
+    add the <literal>append=True</literal> keyword argument.
```
Comment (bdbaddog) [inline doc/user/output.xml]: from this wording I though append and local_only were AddOption() arguments, and not Help() arguments.
Can you rephrase (I guess same comment for CHANGES and RELEASE blurbs)

Comment (mwichmann) [reply doc/user/output.xml]: yeah, my bad, didn't sufficiently rewrite existing text.

File: doc/user/output.xml (hunk 2 above)
Comment (mwichmann) [reply doc/user/output.xml]: I did push a rewrite, see if that's better.

**Comment by bdbaddog** [comment]
Ok. I think I finally understand this. Merging.
